python/insights_ansible_playbook_signer/app.py

```diff
@@ -0,0 +1,179 @@
+        help="Display logs",
+    )
+    keys = parser.add_mutually_exclusive_group(required=True)
+    keys.add_argument(
+        "--key",
+        type=pathlib.Path,
+        help="Path to private GPG key",
+    )
+    keys.add_argument(
+        "--remote-key",
+        type=str,
+        help="Name of a key for a remote signing server",
+    )
+    playbook = parser.add_mutually_exclusive_group(required=True)
+    playbook.add_argument(
+        "--playbook",
+        type=pathlib.Path,
+        help="Path to playbook",
+    )
+    playbook.add_argument(
+        "--stdin",
+        action="store_true",
+        help="Load playbook from stdin (the default)",
+    )
+    args = parser.parse_args()
+
+    # Load playbook to sign
+    raw_playbook: str = ""
+    if args.stdin:
+        with contextlib.suppress(KeyboardInterrupt):
+            raw_playbook = sys.stdin.read()
+    else:
+        raw_playbook = args.playbook.read_text()
+    if len(raw_playbook) == 0:
+        logger.error("Received empty playbook.")
+        raise RuntimeError("Received empty playbook.")
+
+    # Load plays
+    raw_plays: list[dict] = lib.parse_playbook(raw_playbook)
+    if not raw_plays:
+        raise lib.PreconditionError("Playbook contains no plays.")
+    logger.debug(f"Playbook contains {len(raw_plays)} plays.")
+
+    # Sign plays
+    plays: list[dict] = []
+    for i, raw_play in enumerate(raw_plays, 1):
+        play_name: str = raw_play.get("name", "???")
+        logger.debug(f"Preparing to sign play {play_name}.")
+        play: dict = copy.deepcopy(raw_play)
+
+        if "vars" not in play.keys():
+            logger.debug("Filling in missing 'vars' map.")
+            play["vars"] = {}
+        if "insights_signature_exclude" not in play["vars"].keys():
+            logger.debug("Filling in missing 'insights_signature_exclude' pair.")
+            play["vars"]["insights_signature_exclude"] = (
+                "/hosts,/vars/insights_signature"
+            )
+
+        if "insights_signature" not in play["vars"].keys():
+            # The 'clean_play' method requires this to be included.
+            # It will be overwritten later.
+            play["vars"]["insights_signature"] = ""
+
+        # Ensure 'tasks' are the last element
+        if "tasks" in play.keys():
+            play["tasks"] = play.pop("tasks")
+        else:
+            logger.warning("Play does not include 'tasks', is that correct?")
+
+        cleaned_play: dict = lib.clean_play(play)
+        serialized_play: bytes = lib.serialize_play(cleaned_play).encode("utf-8")
+        digest: bytes = lib.create_play_digest(serialized_play)
+
+        logger.debug(f"Serialized play '{play_name}' as {serialized_play!r}")
+
+        signature: bytes
+        if args.remote_key:
+            signature = send_signing_request(digest, key=args.remote_key)
+        else:
+            if not args.key.is_file():
+                raise RuntimeError(f"Key {args.key}' does not exist.")
```

Comment:
In case of `send_signing_request()`, this check is done within the function; for consistence I'd do the same for `sign_play_digest()`, ie perform this existence check in `sign_play_digest()`.